EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs

```diff
@@ -41,34 +42,66 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
                 var typeMapping = (RelationalTypeMapping?)property.FindTypeMapping()
                     ?? _typeMappingSource.FindMapping((IProperty)property);
                 var propertyType = Nullable.GetUnderlyingType(property.ClrType) ?? property.ClrType;
-                if (propertyType.IsEnum
-                    && typeMapping != null
-                    && !propertyType.IsDefined(typeof(FlagsAttribute), true)
-                    && property.GetColumnName(tableIdentifier) is { } columnName)
+                if (!propertyType.IsEnum
+                    || typeMapping is null
+                    || propertyType.IsDefined(typeof(FlagsAttribute), true)
+                    || property.GetColumnName(tableIdentifier) is not { } columnName)
                 {
-                    var enumValues = Enum.GetValues(propertyType);
-                    if (enumValues.Length <= 0)
-                    {
-                        continue;
-                    }
+                    continue;
+                }
+
+                var values = Enum.GetValues(propertyType);
+                if (values.Length == 0)
+                {
+                    continue;
+                }
 
-                    sql.Clear();
+                sql.Clear();
 
-                    sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+                sql.Append(_sqlGenerationHelper.DelimitIdentifier(columnName));
+
+                if (TryParseContiguousRange(values, typeMapping, out var minValue, out var maxValue))
+                {
+                    sql.Append(" BETWEEN ");
+                    sql.Append(minValue);
+                    sql.Append(" AND ");
+                    sql.Append(maxValue);
+                }
+                else
+                {
                     sql.Append(" IN (");
-                    foreach (var item in enumValues)
+                    foreach (var item in values.Cast<object>().Select(typeMapping.GenerateSqlLiteral))
                     {
-                        var value = typeMapping.GenerateSqlLiteral(item);
-                        sql.Append($"{value}, ");
+                        sql.Append(item);
+                        sql.Append(", ");
                     }
 
                     sql.Remove(sql.Length - 2, 2);
-                    sql.Append(")");
-
-                    var constraintName = $"CK_{tableName}_{columnName}_Enum";
-                    entityType.AddCheckConstraint(constraintName, sql.ToString());
+                    sql.Append(')');
                 }
+
+                var constraintName = $"CK_{tableName}_{columnName}_Enum";
+                entityType.AddCheckConstraint(constraintName, sql.ToString());
             }
         }
     }
+
+    private static bool TryParseContiguousRange(IEnumerable values, CoreTypeMapping typeMapping, out decimal minValue, out decimal maxValue)
+    {
+        if (typeMapping.Converter?.ProviderClrType == typeof(string))
```

Comment:
I don't think checking for string is sufficient - there could be a value converter to another non-string non-numeric database type (Byte array? Even timestamp in some convoluted scenario?), and we can't do BETWEEN there. In other words, I think we to do BETWEEN only if we specifically detect a numeric provider CLR types, and refrain from doing it otherwise.

Reply:
I started with a list of types but switched to only string because I thought it was simpler. I have a hard time imagining other converters but you're right, we can't really know all the different conversions available.